pkg/services/sqlstore/migrations/ualert/route_matcher.go

```diff
@@ -0,0 +1,195 @@
+package ualert
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/prometheus/alertmanager/pkg/labels"
+	"xorm.io/xorm"
+)
+
+type migrateRoutes struct {
+	migrator.MigrationBase
+	mg *migrator.Migrator
+}
+
+func (m *migrateRoutes) SQL(dialect migrator.Dialect) string {
+	return "code migration"
+}
+
+func (m *migrateRoutes) Exec(sess *xorm.Session, mg *migrator.Migrator) error {
+	configs := []AlertConfiguration{}
+	err := sess.SQL(`SELECT * FROM alert_configuration`).Find(&configs)
+	if err != nil {
+		return err
+	}
+
+	for i, aCfg := range configs {
+		var oldUserConfig OldPostableUserConfig
+		err = json.Unmarshal([]byte(aCfg.AlertmanagerConfiguration), &oldUserConfig)
+		if err != nil {
+			return fmt.Errorf("unable to read old alertmanager configuration for orgId %v: %w", aCfg.OrgID, err)
+		}
+
+		newRoutePostableApiAlertingConfig := NewRoutePostableApiAlertingConfig{
+			Templates: oldUserConfig.AlertmanagerConfig.Templates,
+			Receivers: oldUserConfig.AlertmanagerConfig.Receivers,
+			Route:     oldRouteToNewRoute(oldUserConfig.AlertmanagerConfig.Route),
+		}
+
+		newConfigUserConfig := NewRoutePostableUserConfig{
+			TemplateFiles:      oldUserConfig.TemplateFiles,
+			AlertmanagerConfig: newRoutePostableApiAlertingConfig,
+		}
+
+		b, err := json.Marshal(newConfigUserConfig)
+		if err != nil {
+			return err
+		}
+
+		configs[i].AlertmanagerConfiguration = string(b)
+	}
+
+	for _, cfg := range configs {
+		_, err = sess.ID(cfg.ID).Update(&cfg)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func oldRouteToNewRoute(topRoute *OldRoute) *NewRoute {
+	if topRoute == nil {
+		return nil
+	}
+	var makeNewRoute func(oR *OldRoute) *NewRoute
+	makeNewRoute = func(oR *OldRoute) *NewRoute {
+		newRoute := &NewRoute{
+			Receiver:       oR.Receiver,
+			GroupByStr:     oR.GroupByStr,
+			Matchers:       oR.Matchers, // Kept for downgrades
+			ObjectMatchers: ObjectMatchers(oR.Matchers),
+			Routes:         make([]*NewRoute, 0, len(oR.Routes)),
+		}
+
+		for _, rt := range oR.Routes {
+			newRoute.Routes = append(newRoute.Routes, makeNewRoute(rt))
+		}
+
+		return newRoute
+	}
+
+	return makeNewRoute(topRoute)
+
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for Matchers.
+func (m *Matchers) UnmarshalJSON(data []byte) error {
+	var lines []string
+	if err := json.Unmarshal(data, &lines); err != nil {
+		return err
+	}
+	for _, line := range lines {
+		pm, err := labels.ParseMatchers(line)
+		if err != nil {
+			return err
+		}
+		*m = append(*m, pm...)
+	}
+	sort.Sort(labels.Matchers(*m))
+	return nil
+}
+
+type OldRoute struct {
```

Comment:
I think I need to add `Continue`, `GroupWait`, `GroupInterval`, and `Repeat Interval` so they get copied in the migration.